### PR TITLE
Add IYK Collab Contract

### PR DESCRIPTION
**Pull request by Ryan-Ouyang**
This PR adds `GenArt721CoreV2_IYK.sol`, an extension of `GenArt721CoreV2_PBAB.sol` with the transfer functions disabled and the IYK claim behavior added.

Goerli deployment of `SignVerifierRegistry`: https://goerli.etherscan.io/address/0xc4aee06cd958a0b489edc0f2bea3d48629315916#code
Default SignVerifier ID: `0x9f28225c7d0ace67fa2516bd7725f3949e9a591de0eae9db822b2cb79f38a6b0`
Default SignVerifier Address: `0xfc95c9ffba80cb60f76c653ea8e5ce01253b6c6a`

Token Name: ITERATION-02
Token Ticker: IT-02
Contract owner: `0x5FE770CF8Ad1DF5DDe2c1447A898B087c275F6c1`

### CODE REVIEW

File: contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol
Comment (jakerockland) [inline contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:13]: Would suggest that it is worth having a documentation comment at the top of this file that clarifies that due to this transfer mechanism, these tokens will not be trade-able via secondary marketplaces.

File: contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol
Comment (jakerockland) [inline contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:24]: Nit/curious: why make this public (rather than private) if also exposing an external `getClaimNonce`?

File: contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol
Comment (jakerockland) [inline contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:108]: Nit: I would encourage making it clear in the `@dev` note that the `claimNonces` value for the `_recipient` is also factored in here in the hash.

File: contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol
Comment (jakerockland) [inline contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:142]: Is it necessary for `claimNonces` to be reset in any way when the registry is updated or is that out-of-scope / not needed?

**Review by jakerockland — APPROVED** [review]
This generally LGTM, but a couple inline suggestions / requests here @Ryan-Ouyang 🙏 

cc @ryley-o as well, would love a review pass from your end also Ryley before merging/shipping.

File: contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol
Comment (Ryan-Ouyang) [reply contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:69]: Yep, there's no need for it for exactly the reason you stated :)

`getClaimSigningHash` pulls the current nonce when `claimNFT` is called, and if you tried to use the same signature then the nonce would already have incremented.

File: contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol
Comment (Ryan-Ouyang) [reply contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:24]: Good q! It's to maintain API parity with our older contracts that have this as `public` instead of `getClaimNonce`.

Will do a quick scan here too since this might not need to be `public` anymore cc @clee681

File: contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol
Comment (Ryan-Ouyang) [reply contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:142]: If the registry is updated and the `signVerifier` changes, all existing signatures are invalidated so I don't see the a security risk if we don't reset `claimNonces`.

If the registry is updated and the `signVerifier` stays the same, then I would assume the intended behaviour is to simply migrate to a new registry without impacting the existing sigs!

So I don't think we need to reset `claimNonces`

Comment (jakerockland) [reply contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:142]: Roger that–SGTM. May be worth clarifying this in a documentation `@dev` note, just for posterity.

File: contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol
Comment (jakerockland) [reply contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:25]: Roger that–SGTM; may be worth clarifying in a `@dev` comment, but this is more of a nit–will defer to y'all.

File: contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol
Comment (jakerockland) [reply contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:70]: Roger that–makes sense to me; may be worth clarifying this inline via a comment, but SGTM 👍 🙏

**Review by ryley-o — CHANGES_REQUESTED** [review]
This is a very cool system!

Going to go ahead and propose some updates here - some are minor nits that you can definitely feel free to ignore! would love to take another look once we process these comments 👍 🙏

File: contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol
Comment (Ryan-Ouyang) [reply contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:25]: > Seems a bit more straightforward and easier for the signVerifier to keep straight?

Could you elaborate on this comment? The `signVerifier` doesn't really need to keep track of any state on its own, it's always fetching from the contract before signing the claiming hash.

Also, keeping it per-user instead of per-token means that the following situation might occur:
1. Person A scans the item and gets a valid signature with token transfer nonce 1
2. Person B scans the item and gets a valid signature with token transfer nonce 1
3. Person A submits the signature to the contract
4. Person B can no longer submit their signature, even though they're the latest possessor of the item.

The above situation is avoided if signatures are submitted immediately but due to fluctuations in gas person A's txn may not go through until later and we don't want to prevent person B from also claiming the NFT.

File: contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol
Comment (Ryan-Ouyang) [reply contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:70]: Nope, nice catch.

File: contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol
Comment (Ryan-Ouyang) [reply contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:152]: Nope, good catch!

File: contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol
Comment (Ryan-Ouyang) [reply contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:113]: Since `claimNonces` gets updated to prevent reusing the current signature, is this necessary? Would prefer to not add the gas cost here unless there's a strong reason to.

Comment (ryley-o) [reply contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:113]: Yeah, I agree that this idea doesn't do anything worthwhile, and just adds gas. Resolving!

File: test/core/GenArt721CoreV2_IYK_Integration.tests.ts
Comment (Ryan-Ouyang) [reply test/core/GenArt721CoreV2_IYK_Integration.tests.ts:465]: Fun thing in Ethers v6 is that we won't need to look up these function signatures as strings anymore!

https://blog.ricmoo.com/highlights-ethers-js-december-2021-dc1adb779d1a

File: contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol
Comment (ryley-o) [reply contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:34]: That is a good point - I was thinking of the following situation based on the current state of this PR:

1. Person A scans the item and gets a valid signature with Address A transfer nonce 1
2. Person B scans the item and gets a valid signature with Address B transfer nonce 1
3. Person B submits the signature to the contract and sees that they own the NFT
4. Person A submits the signature to the contract and now owns the token, while Person B is unaware that they no longer own the token

The remedy to both of our described situations is that person B can re-scan and re-claim their NFT. However, in this example, person B's ownership is custodial (they have to routinely check that Person A has not claimed their token), whereas in the token-transfer-nonce model, person B's ownership is non-custodial. Additionally, in this example, there could be ~infinite addresses between Person A and B (e.g. Person A.1, A.2, A.3, A.4, ...A.`n`), whereas in the token transfer nonce, the denial of service could only happen once.

All of this is likely mitigated by timestamp expiration of the signatures though, so definitely up to you regarding which model is preferred!

**Review by ryley-o — APPROVED** [review]
🙏 🚀 ❤️ LGTM! added a couple nits/minor questions 👍

**Comment by ryley-o** [comment]
@Ryan-Ouyang - our env vars don't populate by default in our CI for forked branches (for security). If you need another round of CI ran, feel free to ping me and I'll queue up the commit 👍 
(also, running a `yarn format` will fix that failing lint test 👍 )

File: contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol
Comment (jakerockland) [reply contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:15]: Yeah probably worth updating this to say "Art Blocks Inc. + IYK" or just "IYK" no strong pref either way 😄

File: contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol
Comment (Ryan-Ouyang) [reply contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:154]: I think since you guys will be deploying this with no plans to do so on other chains, we should be good with this 👍

**Comment by jakerockland** [comment]
Amazing–thank you @Ryan-Ouyang; let us poke a bit to figure out why CI is being angry, but the updates LGTM 🙏 cc @ryley-o

File: contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol
Comment (ryley-o) [reply contracts/PBAB+Collabs/iyk/GenArt721CoreV2_IYK.sol:34]: resolving this 👍

**Pull request merged**
